iree/vm/ops.h

```diff
@@ -127,5 +127,59 @@ static inline int32_t vm_cmp_nz_i32(int32_t operand) {
                                 iree_make_cstring_view("message"));         \
   }
 
+//===------------------------------------------------------------------===//
+// ExtI64: Constants
+//===------------------------------------------------------------------===//
+
+static inline int64_t vm_const_i64(int64_t a) { return a; }
+
+//===------------------------------------------------------------------===//
+// ExtI64: Conditional assignment
+//===------------------------------------------------------------------===//
+
+static inline int64_t vm_select_i64(int32_t condition, int64_t true_value,
+                                    int64_t false_value) {
+  return condition ? true_value : false_value;
+}
+
+//===------------------------------------------------------------------===//
+// ExtI64: Native integer arithmetic ops
+//===------------------------------------------------------------------===//
+
+static inline int64_t vm_add_i64(int64_t lhs, int64_t rhs) { return lhs + rhs; }
+static inline int64_t vm_sub_i64(int64_t lhs, int64_t rhs) { return lhs - rhs; }
+static inline int64_t vm_mul_i64(int64_t lhs, int64_t rhs) { return lhs * rhs; }
+static inline int64_t vm_div_i64s(int64_t lhs, int64_t rhs) {
+  return lhs / rhs;
+}
+static inline int64_t vm_div_i64u(int64_t lhs, int64_t rhs) {
+  return (int64_t)(((uint64_t)lhs) / ((uint64_t)rhs));
+}
+static inline int64_t vm_rem_i64s(int64_t lhs, int64_t rhs) {
+  return lhs % rhs;
+}
+static inline int64_t vm_rem_i64u(int64_t lhs, int64_t rhs) {
+  return (int64_t)(((uint64_t)lhs) % ((uint64_t)rhs));
+}
+static inline int64_t vm_not_i64(int64_t operand) {
+  return (int64_t)(~((uint64_t)operand));
+}
+static inline int64_t vm_and_i64(int64_t lhs, int64_t rhs) { return lhs & rhs; }
+static inline int64_t vm_or_i64(int64_t lhs, int64_t rhs) { return lhs | rhs; }
+static inline int64_t vm_xor_i64(int64_t lhs, int64_t rhs) { return lhs ^ rhs; }
+
+//===------------------------------------------------------------------===//
+// Native bitwise shifts and rotates
+//===------------------------------------------------------------------===//
+
+static inline int64_t vm_shl_i64(int64_t operand, int8_t amount) {
+  return (int64_t)(operand << amount);
```

Comment:
We want to mask these shift amounts to prevent undefined behavior - for example, a shift left of amount=100 will trigger undefined behavior.

from the C spec:
```
  If the value of the right operand is negative, or equal to or greater
  than the number of bits in the promoted left operand, the behavior is
  undefined.
```

Super nice examples: https://wiki.sei.cmu.edu/confluence/display/c/INT34-C.+Do+not+shift+an+expression+by+a+negative+number+of+bits+or+by+greater+than+or+equal+to+the+number+of+bits+that+exist+in+the+operand

We can use uint8_t here for amount so that we don't have to worry about negative shift amounts.
Once we have the unsigned shift amount we can mask `(amount & 63)` (and `(amount & 31)` for the 32-bit variants) and then only need to check for 0 (only perform the shift if `(amount & 63)` != 0).

Something like:
```c
  return (amount & 63) ? (int64_t)(operand << (amount & 63)) : operand;
```

I know the code didn't do this correctly before so if you don't want to fix it here we can file an issue and do it in a followup - my ubsan sense is tingling that we'll end up with fuzzing producing some errors here and we may as well make it right while we remember :)

Tests that shift by 0 and >32/>64 may be hard to author (as I think we check for these in IR and fold them away), but may be worth a shot.

Reply:
Good point! Will fill an issue and we will handle this in a followup.